### ISSUE TITLE
[API Proposal]: SemaphoreSlim.EnterScope

### ISSUE THREAD

**Issue by EgorBo** [dotnet/runtime#103509]
### Background and motivation

A classic pattern for locks in async code looks like this:
```cs
async Task DoWork()
{
    await _semaphoreSlim.WaitAsync();
    try
    {
        // some work
    }
    finally
    {
        _semaphoreSlim.Release();
    }
}
```
This pattern feels a bit too verbose and it's important to keep in mind that `Release` has to be called in `finally` block to avoid dead-locks. I propose we add a new API for `SemaphoreSlim` so the same program can be simplified down to
```cs
async Task DoWork()
{
    await using var _ = await _semaphoreSlim.EnterScope();
    // some work
}
```
Similarily to `Lock.EnterScope` API.

### API Proposal

```diff
namespace System.Threading;

public class SemaphoreSlim
{
+   public Task<SemaphoreSlimReleaser> EnterScope();
+   public Task<SemaphoreSlimReleaser> EnterScope(CancellationToken cancellationToken);
+   public Task<SemaphoreSlimReleaser> EnterScope(int millisecondsTimeout);
+   public Task<SemaphoreSlimReleaser> EnterScope(TimeSpan timeout, CancellationToken cancellationToken);
+   public Task<SemaphoreSlimReleaser> EnterScope(int millisecondsTimeout, CancellationToken cancellationToken);
}

+public sealed class SemaphoreSlimReleaser(SemaphoreSlim semaphore) : IAsyncDisposable
+{
+   public async ValueTask DisposeAsync() => semaphore.Release();
+}
```
I am not sure about the overloads, I just copied the exising ones from `WaitAsync` for consistency, I'd pesonally add only these:
```diff
+   public Task<SemaphoreSlimReleaser> EnterScope(CancellationToken cancellationToken = default);
+   public Task<SemaphoreSlimReleaser> EnterScope(TimeSpan timeout, CancellationToken cancellationToken = default);
```
Also, should it be `EnterScopeAsync` ?

### API Usage

```csharp
await using var _ = await _semaphoreSlim.EnterScope();
```
Sadly, C# doesn't allow to omit the variable so we could have something like:
```cs
await using _semaphoreSlim.EnterScope();
```
Or even extend the `lock` keyword to support `async/await` like we recently extended it to support the new `Lock` type.

**Comment by jcotton42** [comment]
In the case of
```diff
+public sealed class SemaphoreSlimReleaser(SemaphoreSlim semaphore) : IDisposable
+{
+   public void Dispose() => semaphore.Release();
+}
```
Would we maybe want an `idDisposed` bool field or similar, to prevent double release by disposing the same releaser twice?

**Comment by MihaZupan** [comment]
I'd make `EnterScope` a `ValueTask` and change `SemaphoreSlimReleaser` to a `public struct Scope` nested type (similar to `Lock`'s scope in #34812).

**Comment by y0ung3r** [comment]
> Also, should it be `EnterScopeAsync`?

My personal opinion - yes

**Comment by aloraman** [comment]
This pattern of issuing a small *cookie*-object, while changing the state of the owner, and reverting the owner's state on cookie disposal - it's quite familiar and somewhat well-known. I'm so accustomed to it, that I would've chosen the name `SemaphoreReleaseCookie` without a hint of hesitation. That said, MSFT uses the term 'cookie' mostly for stack canary scenarios, so it doesn't fit here.
Nevertheless, I would suggest the following:
 - Do make this cookie object not a class, but a readonly struct.
 - Don't bother with double-release prevention, IMHO, additional security for advanced usage scenarios just isn't worth the overhead.

**Comment by JakeYallop** [comment]
I've used this pattern so many times, I'm surprised something similar hasn't been proposed before. I think the name should be `EnterScopeAsync`, to fit with basically every other API that returns a `Task` or `ValueTask`.

**Comment by Rob-Hague** [comment]
I have also used this pattern a lot and would welcome something more idiomatic. One thing though, is that in practically all cases I am also using the equivalent sync pattern. It would be nice if both async and sync variants of the pattern were available.

By the way, this exists already in Stephen Cleary's package (https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Tasks/SemaphoreSlimExtensions.cs). If we were going there, it would be even nicer if there were a dedicated type which can participate in synchronous and asynchronous mutual exclusion, including with Monitor.Wait&Pulse-like semantics. Basically, something similar to Cleary's AsyncMonitor/AsyncLock (https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncMonitor.cs) but perhaps that's a separate discussion.

**Comment by hez2010** [comment]
imo something better should be used in async code to avoid blocking threads. Like https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-5-asyncsemaphore/

**Comment by lilinus** [comment]
> I'd make `EnterScope` a `ValueTask` and change `SemaphoreSlimReleaser` to a `public struct Scope` nested type (similar to `Lock`'s scope in #34812).

Just to point  out a potential caveat:
If the scope has a field that prevents double-releasing and is a struct, would it not still be possible to release twice if the struct is copied?

**Comment by jaredpar** [comment]
> Similarily to [Lock.EnterScope](https://apisof.net/catalog/19a7f990a1c749e8f0b942b76fd2f7b2) API.

Given the similarity should we consider expanding the new C# `lock` support to include this type? When we started the work on `Lock.EnterScope` in was specific to that type because there were no other examples.

**Comment by EgorBo** [comment]
> > Similarily to [Lock.EnterScope](https://apisof.net/catalog/19a7f990a1c749e8f0b942b76fd2f7b2) API.
> 
> Given the similarity should we consider expanding the new C# `lock` support to include this type? When we started the work on `Lock.EnterScope` in was specific to that type because there were no other examples.

I guess the main problem that it's not obvious what it should do, the wait in this case is async, so should it look like this:
```cs
lock (obj) // await is implicit, requires method to be async. obj must be SemaphoreSlim (or some duck-typing check)
{
}
```
or it needs to be a new construct:
```cs
await lock (obj)
{
}
```

**Comment by quinmars** [comment]
I think it should be `await lock (...)` analog to `using` and `foreach`. More problematic is how you can pass the cancellation token and the timeout and how you can `.ConfigureAwait`?

**Comment by tjwald** [comment]
I like the enter scope as it allows configuring how to enter the scope in a very simple manner.
And I also think that we should also have support for: 
```csharp
await using (var scope = _semaphore.EnterScope(...))
{
    ...
}
```

And this is simply up to the user whether to use the await or the sync version. 
This doesn't require any language changes as this is all supported given that SemaphoreSlim's Scope will implement both `IDisposable` and `IAsyncDisposable`.

If you do choose to support both, there should be an analyzer that tells you to use the await version when in an async functions.

> @quinmars how you can `.ConfigureAwait`? 

How do you configure await on any `IAsyncDisposable`?

**Comment by CaCTuCaTu4ECKuu** [comment]
The thing is **Release** is actually _sync_, but entering scope awaits _async_, so it has to be
```csharp
using var scope = await _semaphore.EnterScopeAsync(cancellationToken);
```
And you should have check for double releasing because you may want to have few locks and manually dispose them in random places

Other thing is that it's not so obvious how to better implement entering lock scope with timeout. It may look like it doesn't matter but when you using time limit you actually want to handle it like this
```csharp
if (!await _semaphore.WaitAsync(timeout))
{
    // Handle fail
    return;
}
```
And with scope it have more variants of implementations that you would like to, from just throwing an exception (which is best for use of scoped locks imo) to making IsLocked property or adding boolean equality comparer to token to check if it has entered lock state (still success lock enter state)

But all of it aside, if you think about it, you actually 80% of time want to just simplify code and write locking of code within method with 1 line and for other cases you actually will be okay with current SemaphoreSlim. But still even covering this single use case I believe justify adding little class/struct and extension method - it's so good where there's a lot of basic locking needed

#### To summarize:
1. `SemaphoreSlimScopeToken` struct
2. Extension method `Task<SemaphoreSlimScopeToken> EnterScopeAsync(this SemaphoreSlim semaphore, CancellationToken ct = default)`
